clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -143,7 +143,22 @@ private boolean tryAppendForSplit(long timestamp, ByteBuffer key, ByteBuffer val
     }
 
     /**
-     * Complete the request.
+     * Abort the batch and complete the future and callbacks.
+     *
+     * @param exception The exception to use to complete the future and awaiting callbacks.
+     */
+    public void abort(RuntimeException exception) {
+        abortRecordAppends();
```

Comment:
Is this safe? It seems like we should document `abort` and `done` as thread-safe and hence we should not invoke `abortRecordAppends`, which is not thread-safe. Or am I missing something?